packages/ai-jsx/src/lib/tts/tts.tsx

```diff
@@ -0,0 +1,360 @@
+ * for us with a text-to-speech service. Typically, this will be a
+ * fetch request to a server that will return an opaque token.
+ */
+export type GetTokenFunction = (provider: string) => Promise<string>;
+
+/**
+ * Defines a base class for text-to-speech services. This class provides
+ * a common interface for text-to-speech services, as well as some basic
+ * infrastructure for playing audio using the HTML5 audio element.
+ */
+export class TextToSpeechBase {
+  protected audio: HTMLAudioElement;
+  /**
+   * The time when the play() method was first called.
+   */
+  protected playMillis: number = 0;
+  /**
+   * The latency between when the play() method is called and when the audio starts playing.
+   */
+  public latency: number = 0.0;
+  /**
+   * Called when the generated audio has started playing.
+   */
+  public onPlaying?: () => void;
+  /**
+   * Called when the generated audio has finished playing out.
+   */
+  public onComplete?: () => void;
+
+  constructor(protected readonly name: string) {
+    this.audio = new Audio();
+    this.audio.onplay = () => console.log(`[${this.name}] tts playing`);
+    this.audio.onpause = () => console.log(`[${this.name}] tts paused`);
+    this.audio.onloadstart = () => console.log(`[${this.name}] tts loadstart`);
+    this.audio.onloadeddata = () => console.log(`[${this.name}] tts loadeddata`);
+    this.audio.oncanplay = () => console.log(`[${this.name}] tts canplay`);
+    this.audio.onplaying = () => {
+      console.log(`[${this.name}] tts playing`);
+      if (this.playMillis) {
+        this.latency = Math.floor(performance.now() - this.playMillis);
+        console.log(`[${this.name}] tts play latency: ${this.latency} ms`);
+      }
+      this.onPlaying?.();
+    };
+  }
+  /**
+   * Converts the given text to speech and plays it using the HTML5 audio element.
+   * This method may be called multiple times, and the audio will be played serially.
+   * Generation may be buffered, use the flush() method to indicate all text has been provided.
+   */
+  play(_text: string) {
+    throw new Error('Method not implemented.');
+  }
+  /**
+   * Flush any buffered text and play the audio.
+   */
+  flush() {}
+  /**
+   * Stop playing any generated audio and discard any buffered text.
+   */
+  stop() {
+    console.log(`[${this.name}] tts stopping`);
+    this.audio.pause();
+    this.cleanUp();
+  }
+  protected cleanUp() {}
+}
+
+/**
+ * Defines a text-to-speech service that requests audio from a server and
+ * plays it in one shot using the HTML5 audio element. The URL to request
+ * audio from the server is constructed using the provided BuildUrlFunction,
+ * allowing this class to be used with a variety of text-to-speech services.
+ */
+export class SimpleTextToSpeech extends TextToSpeechBase {
+  constructor(
+    name: string,
+    protected readonly urlFunc: BuildUrlFunction,
+    private readonly voice: string,
+    private readonly rate: number = 1.0
```

Comment:
```suggestion
    private readonly rate = 1.0
```

This may be unnecessary if you're providing an initial value.